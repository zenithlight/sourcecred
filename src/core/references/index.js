//@flow

export type {URL, ReferenceDetector} from "./referenceDetector";
export {MappedReferenceDetector} from "./mappedReferenceDetector";
export {CascadingReferenceDetector} from "./cascadingReferenceDetector";
export {TranslatingReferenceDetector} from "./translatingReferenceDetector";
