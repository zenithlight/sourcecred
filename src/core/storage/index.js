// @flow

/**
 * Data Storage allows the implementation of a uniform abstraction for I/O
 */
export interface DataStorage {
  /**
   * If the value for the key entered cannot be found, an error should be thrown.
   */
  get(key: string): Promise<Uint8Array>;
}

/**
 * keys should be file-system friendly
 */
export interface WritableDataStorage extends DataStorage {
  set(key: string, value: Uint8Array): Promise<void>;
}
