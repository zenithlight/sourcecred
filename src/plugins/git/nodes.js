// @flow

import deepFreeze from "deep-freeze";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import type {Hash} from "./types";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GIT_PREFIX = NodeAddress.fromParts(["sourcecred", "git"]);
export function _gitAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GIT_PREFIX, ...parts);
}

export const COMMIT_TYPE: "COMMIT" = "COMMIT";

export const Prefix: {|base: NodeAddressT, commit: RawAddress|} = deepFreeze({
  base: GIT_PREFIX,
  commit: _gitAddress(COMMIT_TYPE),
});

export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +hash: Hash,
|};

export type StructuredAddress = CommitAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = NodeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "COMMIT": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: COMMIT_TYPE, hash};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case COMMIT_TYPE:
      return NodeAddress.append(Prefix.commit, x.hash);
    default:
      throw new Error(`Unexpected type ${(x.type: empty)}`);
  }
}
