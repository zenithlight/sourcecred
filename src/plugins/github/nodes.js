// @flow

import deepFreeze from "deep-freeze";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import {botSet} from "./bots";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GITHUB_PREFIX = NodeAddress.fromParts(["sourcecred", "github"]);
export function _githubAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GITHUB_PREFIX, ...parts);
}

export const REPO_TYPE: "REPO" = "REPO";
export const ISSUE_TYPE: "ISSUE" = "ISSUE";
export const PULL_TYPE: "PULL" = "PULL";
export const REVIEW_TYPE: "REVIEW" = "REVIEW";
export const COMMENT_TYPE: "COMMENT" = "COMMENT";
export const COMMIT_TYPE: "COMMIT" = "COMMIT";
export const USERLIKE_TYPE: "USERLIKE" = "USERLIKE";
export const USER_SUBTYPE: "USER" = "USER";
export const BOT_SUBTYPE: "BOT" = "BOT";

export const Prefix: {|
  base: NodeAddressT,
  bot: RawAddress,
  comment: RawAddress,
  commit: RawAddress,
  issue: RawAddress,
  issueComment: RawAddress,
  pull: RawAddress,
  pullComment: RawAddress,
  repo: RawAddress,
  review: RawAddress,
  reviewComment: RawAddress,
  user: RawAddress,
  userlike: RawAddress,
|} = deepFreeze({
  base: GITHUB_PREFIX,
  repo: _githubAddress(REPO_TYPE),
  issue: _githubAddress(ISSUE_TYPE),
  pull: _githubAddress(PULL_TYPE),
  review: _githubAddress(REVIEW_TYPE),
  comment: _githubAddress(COMMENT_TYPE),
  commit: _githubAddress(COMMIT_TYPE),
  userlike: _githubAddress(USERLIKE_TYPE),
  user: _githubAddress(USERLIKE_TYPE, USER_SUBTYPE),
  bot: _githubAddress(USERLIKE_TYPE, BOT_SUBTYPE),
  reviewComment: _githubAddress(COMMENT_TYPE, REVIEW_TYPE),
  issueComment: _githubAddress(COMMENT_TYPE, ISSUE_TYPE),
  pullComment: _githubAddress(COMMENT_TYPE, PULL_TYPE),
});

/**
 * Return the address corresponding to a GitHub login.
 *
 * If the login is considered a bot, then a bot address is returned. Otherwise,
 * a regular user address is returned. The method does not attempt to find out
 * whether the address should actually be an organization address, as we don't
 * yet handle organization addresses.
 *
 * Note: The signature will need to be refactored when we make the list of bots
 * a configuration option rather than a hardcoded constant.
 */
export function loginAddress(username: string): RawAddress {
  const bots = botSet();
  if (bots.has(username)) {
    return NodeAddress.append(Prefix.bot, username);
  } else {
    return NodeAddress.append(Prefix.user, username);
  }
}

export type RepoAddress = {|
  +type: typeof REPO_TYPE,
  +owner: string,
  +name: string,
|};
export type IssueAddress = {|
  +type: typeof ISSUE_TYPE,
  +repo: RepoAddress,
  +number: string,
|};
export type PullAddress = {|
  +type: typeof PULL_TYPE,
  +repo: RepoAddress,
  +number: string,
|};
export type ReviewAddress = {|
  +type: typeof REVIEW_TYPE,
  +pull: PullAddress,
  +id: string,
|};
export type CommentAddress = {|
  +type: typeof COMMENT_TYPE,
  +parent: CommentableAddress,
  +id: string,
|};
export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +id: string,
|};
export type UserlikeAddress = {|
  +type: typeof USERLIKE_TYPE,
  +subtype: typeof USER_SUBTYPE | typeof BOT_SUBTYPE,
  +login: string,
|};

export type StructuredAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress
  | UserlikeAddress;

// Each of these types has 0 or more "AUTHORS" edges, each of which
// leads to a UserlikeAddress.  Note: It is not true that every
// Authorable has at least one author, as when GitHub accounts are
// deleted, they leave behind posts without authors.
export type AuthorableAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress;

// Each of these types has text content, which means
// it may be the source of a reference to a ReferentAddress.
export type TextContentAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress;

// Each of these types may be referred to by something
// with text content.
export type ReferentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress
  | UserlikeAddress;

// Each of these types is structurally the child of some
// entity with a ParentAddress.
export type ChildAddress =
  | IssueAddress // child of RepoAddress
  | PullAddress // child of RepoAddress
  | ReviewAddress // child of PullAddress
  | CommentAddress; // child of IssueAddress, PullAddress, or ReviewAddress

// Each of these types may have child nodes (see ChildAddress).
export type ParentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress;

// GitHub allows you to react to these types
export type ReactableAddress = IssueAddress | PullAddress | CommentAddress;

// Each of these types may have Comments as children
export type CommentableAddress = IssueAddress | PullAddress | ReviewAddress;

// Verify that Commentable is a subtype of Parent
const _unused_static = (_: CommentableAddress): ParentAddress => _;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = NodeAddress.toParts(x);
  switch (kind) {
    case REPO_TYPE: {
      if (rest.length !== 2) {
        throw fail();
      }
      const [owner, name] = rest;
      return {type: REPO_TYPE, owner, name};
    }
    case ISSUE_TYPE: {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      return {type: ISSUE_TYPE, repo, number};
    }
    case PULL_TYPE: {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      return {type: PULL_TYPE, repo, number};
    }
    case REVIEW_TYPE: {
      if (rest.length !== 4) {
        throw fail();
      }
      const [owner, name, pullNumber, id] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      const pull = {type: PULL_TYPE, repo, number: pullNumber};
      return {type: REVIEW_TYPE, pull, id};
    }
    case COMMENT_TYPE: {
      if (rest.length < 1) {
        throw fail();
      }
      const [subkind, ...subrest] = rest;
      switch (subkind) {
        case ISSUE_TYPE: {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, issueNumber, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const issue = {type: ISSUE_TYPE, repo, number: issueNumber};
          return {type: COMMENT_TYPE, parent: issue, id};
        }
        case PULL_TYPE: {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, pullNumber, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const pull = {type: PULL_TYPE, repo, number: pullNumber};
          return {type: COMMENT_TYPE, parent: pull, id};
        }
        case REVIEW_TYPE: {
          if (subrest.length !== 5) {
            throw fail();
          }
          const [owner, name, pullNumber, reviewFragment, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const pull = {type: PULL_TYPE, repo, number: pullNumber};
          const review = {type: REVIEW_TYPE, pull, id: reviewFragment};
          return {type: COMMENT_TYPE, parent: review, id};
        }
        default:
          throw fail();
      }
    }
    case COMMIT_TYPE: {
      if (rest.length !== 1) {
        throw fail();
      }
      const [id] = rest;
      return {type: COMMIT_TYPE, id};
    }
    case USERLIKE_TYPE: {
      if (rest.length !== 2) {
        throw fail();
      }
      const [subtype, login] = rest;
      if (subtype !== "USER" && subtype !== "BOT") {
        throw fail();
      }
      return {type: USERLIKE_TYPE, subtype, login};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case REPO_TYPE:
      return NodeAddress.append(Prefix.repo, x.owner, x.name);
    case ISSUE_TYPE:
      return NodeAddress.append(
        Prefix.issue,
        x.repo.owner,
        x.repo.name,
        x.number
      );
    case PULL_TYPE:
      return NodeAddress.append(
        Prefix.pull,
        x.repo.owner,
        x.repo.name,
        x.number
      );
    case REVIEW_TYPE:
      return NodeAddress.append(
        Prefix.review,
        x.pull.repo.owner,
        x.pull.repo.name,
        x.pull.number,
        x.id
      );
    case COMMENT_TYPE:
      switch (x.parent.type) {
        case ISSUE_TYPE:
          return NodeAddress.append(
            Prefix.issueComment,
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.id
          );
        case PULL_TYPE:
          return NodeAddress.append(
            Prefix.pullComment,
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.id
          );
        case REVIEW_TYPE:
          return NodeAddress.append(
            Prefix.reviewComment,
            x.parent.pull.repo.owner,
            x.parent.pull.repo.name,
            x.parent.pull.number,
            x.parent.id,
            x.id
          );
        default:
          throw new Error(`Bad comment parent type: ${(x.parent.type: empty)}`);
      }
    case USERLIKE_TYPE:
      switch (x.subtype) {
        case "BOT":
          return NodeAddress.append(Prefix.bot, x.login);
        case "USER":
          return NodeAddress.append(Prefix.user, x.login);
        default:
          throw new Error((x.subtype: empty));
      }
    case COMMIT_TYPE:
      return NodeAddress.append(Prefix.commit, x.id);
    default:
      throw new Error(`Unexpected type ${(x.type: empty)}`);
  }
}
