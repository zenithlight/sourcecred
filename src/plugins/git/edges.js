// @flow

import deepFreeze from "deep-freeze";
import {
  type Edge,
  type EdgeAddressT,
  EdgeAddress,
  NodeAddress,
} from "../../core/graph";
import * as GitNode from "./nodes";
import {type TimestampMs} from "../../util/timestamp";

export opaque type RawAddress: EdgeAddressT = EdgeAddressT;

export const HAS_TREE_TYPE: "HAS_TREE" = "HAS_TREE";
export const HAS_PARENT_TYPE: "HAS_PARENT" = "HAS_PARENT";
export const INCLUDES_TYPE: "INCLUDES" = "INCLUDES";
export const BECOMES_TYPE: "BECOMES" = "BECOMES";
export const HAS_CONTENTS_TYPE: "HAS_CONTENTS" = "HAS_CONTENTS";

const GIT_PREFIX = EdgeAddress.fromParts(["sourcecred", "git"]);
function gitEdgeAddress(...parts: string[]): RawAddress {
  return EdgeAddress.append(GIT_PREFIX, ...parts);
}

export const Prefix: {|base: EdgeAddressT, hasParent: RawAddress|} = deepFreeze(
  {
    base: GIT_PREFIX,
    hasParent: gitEdgeAddress(HAS_PARENT_TYPE),
  }
);

export type HasParentAddress = {|
  type: typeof HAS_PARENT_TYPE,
  child: GitNode.CommitAddress,
  parent: GitNode.CommitAddress,
|};

export type StructuredAddress = HasParentAddress;

export const createEdge: {|
  hasParent: (
    child: GitNode.CommitAddress,
    parent: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
|} = deepFreeze({
  hasParent: (
    child: GitNode.CommitAddress,
    parent: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: HAS_PARENT_TYPE, child, parent}),
    src: GitNode.toRaw(child),
    dst: GitNode.toRaw(parent),
    timestampMs,
  }),
});

const NODE_PREFIX_LENGTH = NodeAddress.toParts(GitNode._gitAddress()).length;

function lengthEncode(x: GitNode.RawAddress): $ReadOnlyArray<string> {
  const baseParts = NodeAddress.toParts(x).slice(NODE_PREFIX_LENGTH);
  return [String(baseParts.length), ...baseParts];
}
function lengthDecode(
  x: $ReadOnlyArray<string>,
  fail: () => Error
): {|+parts: $ReadOnlyArray<string>, +rest: $ReadOnlyArray<string>|} {
  if (x.length === 0) {
    // Not length-encoded.
    throw fail();
  }
  const [lengthString, ...allParts] = x;
  const length = parseInt(lengthString, 10);
  if (isNaN(length)) {
    throw fail();
  }
  if (length > allParts.length) {
    // Not enough elements.
    throw fail();
  }
  return {parts: allParts.slice(0, length), rest: allParts.slice(length)};
}
function multiLengthDecode(x: $ReadOnlyArray<string>, fail: () => Error) {
  let remaining = x;
  const partses = [];
  while (remaining.length > 0) {
    const {parts, rest} = lengthDecode(remaining, fail);
    partses.push(parts);
    remaining = rest;
  }
  return partses;
}

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${EdgeAddress.toString(x)}`);
  }
  if (!EdgeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = EdgeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "HAS_PARENT": {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) throw fail();
      const [childParts, parentParts] = parts;
      const child: GitNode.CommitAddress = (GitNode.fromRaw(
        GitNode._gitAddress(...childParts)
      ): any);
      const parent: GitNode.CommitAddress = (GitNode.fromRaw(
        GitNode._gitAddress(...parentParts)
      ): any);
      return {type: HAS_PARENT_TYPE, child, parent};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case HAS_PARENT_TYPE:
      return EdgeAddress.append(
        Prefix.hasParent,
        ...lengthEncode(GitNode.toRaw(x.child)),
        ...lengthEncode(GitNode.toRaw(x.parent))
      );
    default:
      throw new Error((x.type: empty));
  }
}
