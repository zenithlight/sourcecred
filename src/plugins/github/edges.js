// @flow

import deepFreeze from "deep-freeze";
import {
  type Edge,
  type EdgeAddressT,
  EdgeAddress,
  NodeAddress,
} from "../../core/graph";
import * as GithubNode from "./nodes";
import * as GitNode from "../git/nodes";
import {
  type ReactionContent,
  ReactionContent$Values as Reactions,
} from "./graphqlTypes";
import {type TimestampMs} from "../../util/timestamp";

export opaque type RawAddress: EdgeAddressT = EdgeAddressT;

export const AUTHORS_TYPE = "AUTHORS";
export const MERGED_AS_TYPE = "MERGED_AS";
export const HAS_PARENT_TYPE = "HAS_PARENT";
export const REFERENCES_TYPE = "REFERENCES";
export const REACTS_TYPE = "REACTS";
// GitHub tracks its own notion of a commit, which has a particular
// database id, is scoped to a particular repository, and has a canonical url
// on GitHub. The CORRESPONDS_TO_COMMIT_TYPE edges connect the GitHub commits
// to the corresponding Git commit.
export const CORRESPONDS_TO_COMMIT_TYPE = "CORRESPONDS_TO_COMMIT_TYPE";

const GITHUB_PREFIX = EdgeAddress.fromParts(["sourcecred", "github"]);
function githubEdgeAddress(...parts: string[]): RawAddress {
  return EdgeAddress.append(GITHUB_PREFIX, ...parts);
}

export const Prefix: {|
  authors: RawAddress,
  base: EdgeAddressT,
  correspondsToCommit: RawAddress,
  hasParent: RawAddress,
  mergedAs: RawAddress,
  reacts: RawAddress,
  reactsHeart: RawAddress,
  reactsHooray: RawAddress,
  reactsRocket: RawAddress,
  reactsThumbsUp: RawAddress,
  references: RawAddress,
|} = deepFreeze({
  base: GITHUB_PREFIX,
  authors: githubEdgeAddress(AUTHORS_TYPE),
  mergedAs: githubEdgeAddress(MERGED_AS_TYPE),
  references: githubEdgeAddress(REFERENCES_TYPE),
  hasParent: githubEdgeAddress(HAS_PARENT_TYPE),
  reacts: githubEdgeAddress(REACTS_TYPE),
  reactsThumbsUp: githubEdgeAddress(REACTS_TYPE, Reactions.THUMBS_UP),
  reactsHeart: githubEdgeAddress(REACTS_TYPE, Reactions.HEART),
  reactsHooray: githubEdgeAddress(REACTS_TYPE, Reactions.HOORAY),
  reactsRocket: githubEdgeAddress(REACTS_TYPE, Reactions.ROCKET),
  correspondsToCommit: githubEdgeAddress(CORRESPONDS_TO_COMMIT_TYPE),
});

export type AuthorsAddress = {|
  +type: typeof AUTHORS_TYPE,
  +author: GithubNode.UserlikeAddress,
  +content: GithubNode.AuthorableAddress,
|};
export type MergedAsAddress = {|
  +type: typeof MERGED_AS_TYPE,
  +pull: GithubNode.PullAddress,
|};
export type HasParentAddress = {|
  +type: typeof HAS_PARENT_TYPE,
  +child: GithubNode.ChildAddress,
|};
export type ReferencesAddress = {|
  +type: typeof REFERENCES_TYPE,
  +referrer: GithubNode.TextContentAddress,
  +referent: GithubNode.ReferentAddress,
|};
export type ReactsAddress = {|
  +type: typeof REACTS_TYPE,
  +reactionType: ReactionContent,
  +user: GithubNode.UserlikeAddress,
  +reactable: GithubNode.ReactableAddress,
|};
export type CorrespondsToCommitAddress = {|
  +type: typeof CORRESPONDS_TO_COMMIT_TYPE,
  +githubCommit: GithubNode.CommitAddress,
|};

export type StructuredAddress =
  | AuthorsAddress
  | MergedAsAddress
  | HasParentAddress
  | ReferencesAddress
  | ReactsAddress
  | CorrespondsToCommitAddress;

export const createEdge: {|
  authors: (
    author: GithubNode.UserlikeAddress,
    content: GithubNode.AuthorableAddress,
    timestampMs: TimestampMs
  ) => Edge,
  correspondsToCommit: (
    githubCommit: GithubNode.CommitAddress,
    gitCommit: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
  hasParent: (
    child: GithubNode.ChildAddress,
    parent: GithubNode.ParentAddress,
    timestampMs: TimestampMs
  ) => Edge,
  mergedAs: (
    pull: GithubNode.PullAddress,
    commit: GithubNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
  reacts: (
    reactionType: ReactionContent,
    user: GithubNode.UserlikeAddress,
    reactable: GithubNode.ReactableAddress,
    timestampMs: TimestampMs
  ) => Edge,
  references: (
    referrer: GithubNode.TextContentAddress,
    referent: GithubNode.ReferentAddress,
    timestampMs: TimestampMs
  ) => Edge,
|} = deepFreeze({
  authors: (
    author: GithubNode.UserlikeAddress,
    content: GithubNode.AuthorableAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: AUTHORS_TYPE, author, content}),
    src: GithubNode.toRaw(author),
    dst: GithubNode.toRaw(content),
    timestampMs,
  }),
  mergedAs: (
    pull: GithubNode.PullAddress,
    commit: GithubNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: MERGED_AS_TYPE, pull}),
    src: GithubNode.toRaw(pull),
    dst: GithubNode.toRaw(commit),
    timestampMs,
  }),
  correspondsToCommit: (
    githubCommit: GithubNode.CommitAddress,
    gitCommit: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: CORRESPONDS_TO_COMMIT_TYPE, githubCommit}),
    src: GithubNode.toRaw(githubCommit),
    dst: GitNode.toRaw(gitCommit),
    timestampMs,
  }),
  hasParent: (
    child: GithubNode.ChildAddress,
    parent: GithubNode.ParentAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: HAS_PARENT_TYPE, child}),
    src: GithubNode.toRaw(child),
    dst: GithubNode.toRaw(parent),
    timestampMs,
  }),
  references: (
    referrer: GithubNode.TextContentAddress,
    referent: GithubNode.ReferentAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: REFERENCES_TYPE, referrer, referent}),
    src: GithubNode.toRaw(referrer),
    dst: GithubNode.toRaw(referent),
    timestampMs,
  }),
  reacts: (
    reactionType: ReactionContent,
    user: GithubNode.UserlikeAddress,
    reactable: GithubNode.ReactableAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({
      type: REACTS_TYPE,
      user,
      reactionType,
      reactable,
    }),
    src: GithubNode.toRaw(user),
    dst: GithubNode.toRaw(reactable),
    timestampMs,
  }),
});

function lengthEncode(x: GithubNode.RawAddress): $ReadOnlyArray<string> {
  const parts = NodeAddress.toParts(x);
  return [String(parts.length), ...parts];
}
function lengthDecode(
  x: $ReadOnlyArray<string>,
  fail: () => Error
): {|+parts: $ReadOnlyArray<string>, +rest: $ReadOnlyArray<string>|} {
  if (x.length === 0) {
    // Not length-encoded.
    throw fail();
  }
  const [lengthString, ...allParts] = x;
  const length = parseInt(lengthString, 10);
  if (isNaN(length)) {
    throw fail();
  }
  if (length > allParts.length) {
    // Not enough elements.
    throw fail();
  }
  return {parts: allParts.slice(0, length), rest: allParts.slice(length)};
}
function multiLengthDecode(x: $ReadOnlyArray<string>, fail: () => Error) {
  let remaining = x;
  const partses = [];
  while (remaining.length > 0) {
    const {parts, rest} = lengthDecode(remaining, fail);
    partses.push(parts);
    remaining = rest;
  }
  return partses;
}

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${EdgeAddress.toString(x)}`);
  }
  if (!EdgeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = EdgeAddress.toParts(x);
  switch (kind) {
    case AUTHORS_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [authorParts, contentParts] = parts;
      const author: GithubNode.UserlikeAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(authorParts): any)
      ): any);
      const content: GithubNode.AuthorableAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(contentParts): any)
      ): any);
      return ({type: AUTHORS_TYPE, author, content}: AuthorsAddress);
    }
    case MERGED_AS_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 1) {
        throw fail();
      }
      const [pullParts] = parts;
      const pull: GithubNode.PullAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(pullParts): any)
      ): any);
      return ({type: MERGED_AS_TYPE, pull}: MergedAsAddress);
    }
    case HAS_PARENT_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 1) {
        throw fail();
      }
      const [childParts] = parts;
      const child: GithubNode.ChildAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(childParts): any)
      ): any);
      return ({type: HAS_PARENT_TYPE, child}: HasParentAddress);
    }
    case REFERENCES_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [referrerParts, referentParts] = parts;
      const referrer: GithubNode.TextContentAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(referrerParts): any)
      ): any);
      const referent: GithubNode.ReferentAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(referentParts): any)
      ): any);
      return ({type: REFERENCES_TYPE, referrer, referent}: ReferencesAddress);
    }
    case REACTS_TYPE: {
      const [rawReactionType, ...rest2] = rest;
      const reactionType = Reactions[rawReactionType];
      if (reactionType == null) {
        throw fail();
      }
      const parts = multiLengthDecode(rest2, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [userParts, reactableParts] = parts;
      const user: GithubNode.UserlikeAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(userParts): any)
      ): any);
      const reactable: GithubNode.ReactableAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(reactableParts): any)
      ): any);
      return {type: REACTS_TYPE, reactionType, user, reactable};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case AUTHORS_TYPE:
      return EdgeAddress.append(
        Prefix.authors,
        ...lengthEncode(GithubNode.toRaw(x.author)),
        ...lengthEncode(GithubNode.toRaw(x.content))
      );
    case MERGED_AS_TYPE:
      return EdgeAddress.append(
        Prefix.mergedAs,
        ...lengthEncode(GithubNode.toRaw(x.pull))
      );
    case HAS_PARENT_TYPE:
      return EdgeAddress.append(
        Prefix.hasParent,
        ...lengthEncode(GithubNode.toRaw(x.child))
      );
    case REFERENCES_TYPE:
      return EdgeAddress.append(
        Prefix.references,
        ...lengthEncode(GithubNode.toRaw(x.referrer)),
        ...lengthEncode(GithubNode.toRaw(x.referent))
      );
    case REACTS_TYPE:
      return EdgeAddress.append(
        Prefix.reacts,
        x.reactionType,
        ...lengthEncode(GithubNode.toRaw(x.user)),
        ...lengthEncode(GithubNode.toRaw(x.reactable))
      );
    case CORRESPONDS_TO_COMMIT_TYPE:
      return EdgeAddress.append(
        Prefix.correspondsToCommit,
        ...lengthEncode(GithubNode.toRaw(x.githubCommit))
      );
    default:
      throw new Error((x.type: empty));
  }
}
