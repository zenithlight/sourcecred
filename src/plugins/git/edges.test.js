// @flow

import {type EdgeAddressT, edgeToParts} from "../../core/graph";
import {createEdge, fromRaw, toRaw} from "./edges";
import * as GE from "./edges";
import * as GN from "./nodes";

describe("plugins/git/edges", () => {
  const nodeExamples = {
    commit: (): GN.CommitAddress => ({
      type: GN.COMMIT_TYPE,
      hash: "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
    }),
    parentCommit: (): GN.CommitAddress => ({
      type: GN.COMMIT_TYPE,
      hash: "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
    }),
  };

  const edgeExamples = {
    hasParent: () =>
      createEdge.hasParent(
        nodeExamples.commit(),
        nodeExamples.parentCommit(),
        102
      ),
  };

  describe("createEdge", () => {
    Object.keys(edgeExamples).forEach((example) => {
      it(`works for ${JSON.stringify(example)}`, () => {
        const instance = edgeExamples[example]();
        expect(edgeToParts(instance)).toMatchSnapshot();
      });
    });
  });

  describe("`toRaw` after `fromRaw` is identity", () => {
    Object.keys(edgeExamples).forEach((example) => {
      it(example, () => {
        const baseAddress: EdgeAddressT = edgeExamples[example]().address;
        const instance: GE.RawAddress = (baseAddress: any);
        expect(toRaw(fromRaw(instance))).toEqual(instance);
      });
    });
  });

  describe("`fromRaw` after `toRaw` is identity", () => {
    Object.keys(edgeExamples).forEach((example) => {
      it(example, () => {
        const baseAddress: EdgeAddressT = edgeExamples[example]().address;
        const instance: GE.RawAddress = (baseAddress: any);
        const structured: GE.StructuredAddress = fromRaw(instance);
        expect(fromRaw(toRaw(structured))).toEqual(structured);
      });
    });
  });
});
