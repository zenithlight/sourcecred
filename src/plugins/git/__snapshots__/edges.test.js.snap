// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/git/edges createEdge works for "hasParent" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "git",
    "HAS_PARENT",
    "2",
    "COMMIT",
    "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
    "2",
    "COMMIT",
    "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
  ],
  "dstParts": Array [
    "sourcecred",
    "git",
    "COMMIT",
    "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
  ],
  "srcParts": Array [
    "sourcecred",
    "git",
    "COMMIT",
    "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
  ],
  "timestampMs": 102,
}
`;
